pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchMessageContainerImpl.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import io.netty.buffer.ByteBuf;
+import java.nio.ByteBuffer;
+import java.util.Set;
+import org.apache.pulsar.client.api.CryptoKeyReader;
+import org.apache.pulsar.client.api.MessageCrypto;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.impl.crypto.MessageCryptoBc;
+import org.apache.pulsar.common.allocator.PulsarByteBufAllocator;
+import org.apache.pulsar.common.api.EncryptionContext;
+import org.apache.pulsar.common.api.proto.CompressionType;
+import org.apache.pulsar.common.api.proto.MessageIdData;
+import org.apache.pulsar.common.api.proto.MessageMetadata;
+import org.apache.pulsar.common.compression.CompressionCodecNone;
+import org.apache.pulsar.common.compression.CompressionCodecProvider;
+import org.apache.pulsar.common.protocol.Commands;
+
+/**
+ * A raw batch message container without producer. (Used for StrategicTwoPhaseCompactor)
+ *
+ * incoming single messages:
+ * (k1, v1), (k2, v1), (k3, v1), (k1, v2), (k2, v2), (k3, v2), (k1, v3), (k2, v3), (k3, v3)
+ *
+ * batched into single batch message:
+ * [(k1, v1), (k2, v1), (k3, v1), (k1, v2), (k2, v2), (k3, v2), (k1, v3), (k2, v3), (k3, v3)]
+ */
+public class RawBatchMessageContainerImpl extends BatchMessageContainerImpl {
```

Comment:
It seems that this is just a need for a bulk message container based on the maximum number of batches, and I feel that extending the `BatchMessageContainerImpl` will become too heavy(There are various producer-related properties and operations in the BatchMessageContainer implementation), and it also adds to the complexity of the original BatchMessageContainer.

Maybe a simple batch container could be re-implemented. What do you think?

Reply:
We want to reuse the batching and payload serialization logic from the parent class.

I do not think this PR adds significant complexity to the parent class.
It just checks if the producer is null or not.

We could pass a mock producer if adding the producer null-check is not desirable.

Reply:
oh, sorry, I miss some code. 

> We could pass a mock producer if adding the producer null-check is not desirable.

I'm leaning towards this.

other, I wonder why batch processing is needed here? Is it because it is known that there is a performance bottleneck here?

Reply:
My main concern is that developers will miss this later when they modify it. Using mock producer is transfer behavior to `RawBatchMessageContainerImpl `.

I revisited the code, and the current implementation is not quite capable of the mock producer. 

like this:

https://github.com/apache/pulsar/blob/4faed760376d6c5bec8803f97c4ed2452664ac69/pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java#L116-L117

https://github.com/apache/pulsar/blob/4faed760376d6c5bec8803f97c4ed2452664ac69/pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java#L136-L138

BWT: In the original implementation, `BatchContainer` coupled with the producer I didn't think was very good. We should make `BatchContiner` independent, Leave things like `releaseMemory` to the producer. (This is off topic and does not affect this PR)